src/nvapi/nvapi_d3d_low_latency_device.cpp

```diff
@@ -1,145 +1,65 @@
-
-    uint64_t LowLatencyFrameIdGenerator::GetLowLatencyDeviceFrameId(uint64_t applicationFrameId) {
-        std::scoped_lock lock(m_frameIdGeneratorMutex);
-
-        auto it = m_applicationIdToDeviceId.find(applicationFrameId);
-        if (it != m_applicationIdToDeviceId.end())
-            return it->second;
-
-        uint64_t lowLatencyDeviceFrameId = m_nextLowLatencyDeviceFrameId++;
-        uint64_t frameIdIndex = (lowLatencyDeviceFrameId - 1) % applicationIdListSize;
-
-        if ((lowLatencyDeviceFrameId - 1) >= applicationIdListSize)
-            m_applicationIdToDeviceId.erase(m_applicationIdList[frameIdIndex]);
-
-        m_applicationIdToDeviceId[applicationFrameId] = lowLatencyDeviceFrameId;
-        m_applicationIdList[frameIdIndex] = applicationFrameId;
-
-        return lowLatencyDeviceFrameId;
-    }
-
-    bool LowLatencyFrameIdGenerator::LowLatencyDeviceFrameIdInWindow(uint64_t lowLatencyDeviceFrameId) const {
-        return ((lowLatencyDeviceFrameId < m_nextLowLatencyDeviceFrameId)
-            && ((m_nextLowLatencyDeviceFrameId - lowLatencyDeviceFrameId) < applicationIdListSize));
-    }
-
-    uint64_t LowLatencyFrameIdGenerator::GetApplicationFrameId(uint64_t lowLatencyDeviceFrameId) {
-        std::scoped_lock lock(m_frameIdGeneratorMutex);
+    std::unordered_map<IUnknown*, std::shared_ptr<NvapiD3dLowLatencyDevice>> NvapiD3dLowLatencyDevice::m_lowLatencyDeviceMap = {};
+    std::mutex NvapiD3dLowLatencyDevice::m_mutex = {};
 
-        if (!lowLatencyDeviceFrameId || !LowLatencyDeviceFrameIdInWindow(lowLatencyDeviceFrameId))
-            return 0;
-
-        return m_applicationIdList[((lowLatencyDeviceFrameId - 1) % applicationIdListSize)];
-    }
-
-    bool LowLatencyFrameIdGenerator::IsRepeatedFrame(uint64_t frameID, uint32_t markerType) {
-        // Should always be within bounds since we drop unsupported marker types in the entrypoints
-        return std::exchange(m_lastFrameId[markerType], frameID) == frameID;
-    }
-
-    bool NvapiD3dLowLatencyDevice::SupportsLowLatency(IUnknown* device) {
-        auto d3dLowLatencyDevice = GetLowLatencyDevice(device);
-        if (d3dLowLatencyDevice == nullptr)
-            return false;
-
-        return d3dLowLatencyDevice->SupportsLowLatency();
-    }
-
-    bool NvapiD3dLowLatencyDevice::SupportsLowLatency(ID3D12CommandQueue* commandQueue) {
-        auto d3dLowLatencyDevice = GetLowLatencyDevice(commandQueue);
-        if (d3dLowLatencyDevice == nullptr)
-            return false;
-
-        return d3dLowLatencyDevice->SupportsLowLatency();
+    void NvapiD3dLowLatencyDevice::Reset() {
+        std::scoped_lock lock{m_mutex};
+        m_lowLatencyDeviceMap.clear();
     }
 
-    bool NvapiD3dLowLatencyDevice::LatencySleep(IUnknown* device) {
-        auto d3dLowLatencyDevice = GetLowLatencyDevice(device);
-        if (d3dLowLatencyDevice == nullptr)
-            return false;
+    static Com<ID3DLowLatencyDevice> GetD3DLowLatencyDevice(IUnknown* device) {
```

Comment:
Maybe rework this as something like…

```suggestion
    static std::pair<IUnknown*, ID3DLowLatencyDevice*> GetDevices(IUnknown* unknown) {
```

… where the first returned interface would be `ID3D11Device` or `ID3D12Device` to be used as "normalized" map key? This way we could search the map for existing device entry with simple `find` and it it wasn't found (and the result is different from `IUnknown*` passed as parameter), we could add both entries ahead of time.

Reply:
Thanks. I've kept the original version for now, I find it slightly easier to read.